opt: generate better constrained index scans using partition values

ridwanmsharif opened the pull request:
Fixes #38031 

This change leverages the partitioning information of an index
to generate more constrained index scans using indexes that would
otherwise be ignored. It assumes that values are likely in a
partition if the partition has been defined but also adds the
in-between spans to guarantee correctness. It tries to isolate
the partition spans for maximum benefit.

Release note: None

Commit:
opt: generate better constrained index scans using partition values

Fixes #38031.

This change leverages the partitioning information of an index
to generate more constrained index scans using indexes that would
otherwise be ignored. It assumes that values are likely in a
partition if the partition has been defined but also adds the
in-between spans to guarantee correctness. It tries to isolate
the partition spans for maximum benefit.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/38963)
<!-- Reviewable:end -->

RaduBerinde reviewed (commented):
Thanks for working on this! This is turning out to involve some interesting problems.

It would be great if we could find a way to avoid adding the "isolated" concept at the level of the Span which seems fragile. The way I was thinking was that we would return only the "likely" spans but store some extra information on the side (maybe in the `indexConstraintCtx`), and incorporate the unlikely spans at the very end (at which point, they won't be merged if the likely spans were extended). It would be worth investigating something like that.

CC @andy-kimball for more ideas.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-:-LkH8eipDrYpOS_GK7zm:bwp4rgb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ridwanmsharif)

___
*[pkg/sql/opt/ops/scalar.opt, line 315 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkH9wkA7Tma87LdNBbv:-LkH9wkA7Tma87LdNBbw:b-i52ha2) ([raw file](https://github.com/cockroachdb/cockroach/blob/45cf0005c1c83509f3653d00d94a27a35ad5c7d6/pkg/sql/opt/ops/scalar.opt#L315)):*
> ```
> # being - the spans created by the index constraint generator will try to keep
> # the Likely spans isolated. That is, the spans created by the Likely expressions
> # will not be merged or consolidated with the Unlikely spans.
> ```

Spans are a more particular concept than what these files should be discussing. It's not even clear which spans we are talking about (we have index constraints and scalar constraints). I would distill it into something more general - "the disjuncts are organized into two categories: the "likely" category and the "unlikely" category; an input is likely to satisfy at least one of the "likely" disjuncts. This categorization is used to incorporate likely column values into index constraint generation."
___
*[pkg/sql/opt/ops/scalar.opt, line 320 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkH9TTI0ghBaXME3cT8:-LkH9TTI0ghBaXME3cT9:b-ymv1k4) ([raw file](https://github.com/cockroachdb/cockroach/blob/45cf0005c1c83509f3653d00d94a27a35ad5c7d6/pkg/sql/opt/ops/scalar.opt#L320)):*
> ```
> [Scalar, Bool]
> define Likely {
>     # Conditions that are likely satisfied by rows.
> ```

This isn't a clear definition - we can't say each is likely satisfied; we can say it's likely that one of them is satisfied.

Naming and comments aside, I like this as a concept, especially the fact that it's semantically equivalent to Or.


<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
I added the isolated concept partly because it helped me find bugs in the spans created by this change and because I thought it was easier to see the partition spans in the `EXPLAIN` plans. I agree now that it looks a little fragile and is sometimes confusing to see unconsolidated plans. Its removed now. I still however like the idea of, passing down to the index constraint library the set of in between filters, so it doesn't have to figure out based on the index and the partition values, what other spans to create. The way we're doing it now adds only a little bit of additional logic to the index constraint library and allows it to reuse everything else.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-:-LkQNkqb9HIoUbdyIhPT:b-uj2odb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/opt/ops/scalar.opt, line 315 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkH9wkA7Tma87LdNBbv:-LkQNOQZ7_A2uZmhHECb:bfk8vd5) ([raw file](https://github.com/cockroachdb/cockroach/blob/45cf0005c1c83509f3653d00d94a27a35ad5c7d6/pkg/sql/opt/ops/scalar.opt#L315)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Spans are a more particular concept than what these files should be discussing. It's not even clear which spans we are talking about (we have index constraints and scalar constraints). I would distill it into something more general - "the disjuncts are organized into two categories: the "likely" category and the "unlikely" category; an input is likely to satisfy at least one of the "likely" disjuncts. This categorization is used to incorporate likely column values into index constraint generation."
</blockquote></details>

Made it a little clearer. Let me know if this is better.
___
*[pkg/sql/opt/ops/scalar.opt, line 320 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkH9TTI0ghBaXME3cT8:-LkQNUx37OX-i9EUPTuE:bol9tm8) ([raw file](https://github.com/cockroachdb/cockroach/blob/45cf0005c1c83509f3653d00d94a27a35ad5c7d6/pkg/sql/opt/ops/scalar.opt#L320)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This isn't a clear definition - we can't say each is likely satisfied; we can say it's likely that one of them is satisfied.

Naming and comments aside, I like this as a concept, especially the fact that it's semantically equivalent to Or.
</blockquote></details>

Ack. Naming is hard. But I like the idea that we generate the two kinds of spans that won't merge together here, it lets us reuse a lot of the index constraint logic as is.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38963)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj, @RaduBerinde, @ridwanmsharif, and @rytaft)

___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_constrained_scans, line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkWbsEH1DX5_kHMwYs5:-LkWbsEH1DX5_kHMwYs6:b-t5c4c3) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/ccl/logictestccl/testdata/logic_test/partitioning_constrained_scans#L53)):*
> ```
>     planet STRING, region STRING, subregion STRING, val INT8,
>     PRIMARY KEY (planet, region, subregion, val)
>     )
> ```

indentation seems off in this query
___
*[pkg/sql/opt/constraint/constraint.go, line 81 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkWcIGp7UEz53i4EN1c:-LkWcIGp7UEz53i4EN1d:b-7ra5hm) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/constraint/constraint.go#L81)):*
> ```Go
> // with the current constraint. It is useful for scans that involve partitions
> // as it lets you see the partition spans clearly.
> func (c *Constraint) UnmergedUnionWith(evalCtx *tree.EvalContext, other *Constraint) {
> ```

This requires some clarification, and it needs unit tests (like `UnionWith` has). What does it mean exactly that they are not merged? It's one thing to not merge spans that "touch" like `[/a -  /b) [/b - /c]` and another thing to not merge spans that actually overlap (like `[/a - /c] [/b - /d]`. The spans in a Constraint are non-overlapping and if we violate that, a lot of code can break..


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Nice work!  I haven't been part of discussions about this so may be missing something, but I don't really understand the need for the `Likely` data type. Radu's idea about adding the unlikely spans at the end makes sense to me....

> Reviewed 8 of 8 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-:-LkZ8kIE1kfEwQKIdQgV:blg82wp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj and @ridwanmsharif)

___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_constrained_scans, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkYh1qDCRpPMgRSfo_D:-LkYh1qDCRpPMgRSfo_E:bggt4f6) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/ccl/logictestccl/testdata/logic_test/partitioning_constrained_scans#L1)):*
> ```
> # LogicTest: local-opt
> ```

should you add `fakedist-opt` or one of the others?
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 741 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkYypcY22ijJgeFLdD6:-LkYypcY22ijJgeFLdD7:b-687kw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/idxconstraint/index_constraints.go#L741)):*
> ```Go
> }
> 
> // makeSpansForLikelySatisfied is similar to Or except it
> ```

[nit] similar to `makeSpansForOr`
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 780 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkYzPiKEi883T5B2ZAg:-LkYzPiKEi883T5B2ZAh:b-fdce6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/idxconstraint/index_constraints.go#L780)):*
> ```Go
> 
> 	// We use an unmerged union so its easier to see the partition spans
> 	// in the EXPLAIN plans. It shouldn't have an effect on anything else.
> ```

Is this really necessary? What does the plan look like when they are merged?
___
*[pkg/sql/opt/ops/scalar.opt, line 311 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ6yw6ESzVMpbw6BTv:-LkZ6yw6ESzVMpbw6BTw:b-jbdkkd) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/ops/scalar.opt#L311)):*
> ```
> }
> 
> # Likely is logically equivalent to an Or expression across all the scalar
> ```

Having a separate type here for the sole purpose of not merging spans seems like a lot of (possibly unnecessary?) overhead
___
*[pkg/sql/opt/ops/scalar.opt, line 317 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ-1KpDNkh4rI2lq8L:-LkZ-1KpDNkh4rI2lq8M:b7ivvg3) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/ops/scalar.opt#L317)):*
> ```
> # and the LikelyNotSatisfied categories; an input is likely to satisfy at least
> # one of the likely expressions.
> # This categorization is used to incorporate likely column values into index
> ```

[super nit] kind of weird that this is on a new line but not a new paragraph.
___
*[pkg/sql/opt/xform/custom_funcs.go, line 206 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ-yA4-dfoLIjdOTU9:-LkZ-yA4-dfoLIjdOTUA:b-uw1lfg) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/custom_funcs.go#L206)):*
> ```Go
> 
> 		if comp == 0 {
> 			filterExpr = append(filterExpr, memo.FiltersItem{Condition: c.e.f.ConstructEq(idxCol, scalarVal)})
> ```

[nit] could reduce some clutter here by assigning the comparison expr to a variable in each case (since that's the only thing that changes), and then create the `FiltersItem` and append it to `filterExpr` below.
___
*[pkg/sql/opt/xform/custom_funcs.go, line 225 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ0jq6CvwKa1NqYWoP:-LkZ0jq6CvwKa1NqYWoQ:bxj92e3) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/custom_funcs.go#L225)):*
> ```Go
> 
> // isPrefixOf returns whether small is a prefix of large.
> func (c *CustomFuncs) isPrefixOf(small []tree.Datum, large []tree.Datum) bool {
> ```

Can you guarantee that this will never cause an array out of bounds error? Needs a comment explaining this.
___
*[pkg/sql/opt/xform/custom_funcs.go, line 237 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ1LpW44q02WwgO0Ne:-LkZ1LpW44q02WwgO0Nf:b6p15pd) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/custom_funcs.go#L237)):*
> ```Go
> // inBetweenFilters returns a set of filters that are required to cover all the
> // in-between spans given a set of partition values. This is required for
> // correctness reasons, although values are unlikely to exist between defined
> ```

[nit] change this comma to semi-colon to fix the run-on sentence
___
*[pkg/sql/opt/xform/custom_funcs.go, line 262 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ1pV8CWWLq6jWhnQG:-LkZ1pV8CWWLq6jWhnQH:b4db0eh) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/custom_funcs.go#L262)):*
> ```Go
> 	inBetween = append(inBetween, &endExpr)
> 
> 	if len(partitionValues) == 1 {
> ```

you don't need this check since you'll just exit the loop below before any iterations
___
*[pkg/sql/opt/xform/custom_funcs.go, line 311 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ5vn9Di5-a2k9q5vH:-LkZ5vn9Di5-a2k9q5vI:b-byar4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/custom_funcs.go#L311)):*
> ```Go
> 
> 	// Construct all the partition filters.
> 	for i, partitionList := range partitionValues {
> ```

[nit] `partition` seems more appropriate than `partitionList`
___
*[pkg/sql/opt/xform/custom_funcs.go, line 316 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ5Dmk2iPNTvtaGMf9:-LkZ5Dmk2iPNTvtaGMfA:b-kexf7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/custom_funcs.go#L316)):*
> ```Go
> 		// been defined on the same partition.
> 		partitionSeen := false
> 		for j, seenPartition := range partitionValues {
> ```

[nit] don't love the naming here (`partitionSeen` v. `seenPartition`). Can you find a way to distinguish these better?
___
*[pkg/sql/opt/xform/custom_funcs.go, line 508 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ7YVi48-5qLt_hwco:-LkZ7YVj9FiJVVchW9Zd:b5lqmk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/custom_funcs.go#L508)):*
> ```Go
> 		}
> 
> 		// If a check constraint filter wasn't able to constrain the index, it
> ```

Update comment to mention partitioned index
___
*[pkg/sql/opt/xform/testdata/rules/partitioned, line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ7pvY0k59tGT1-kUt:-LkZ7pvY0k59tGT1-kUu:baicf6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/testdata/rules/partitioned#L7)):*
> ```
> # Create a truly distributed table. Partitioned appropriately.
> exec-ddl
> CREATE TABLE partitioned (
> ```

what about adding another test with some regions/subregions in mars and/or Jupiter as well?
___
*[pkg/sql/opt/xform/testdata/rules/partitioned, line 16 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ7i6o0kfqIhpgzlJZ:-LkZ7i6o0kfqIhpgzlJ_:bliamkg) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/testdata/rules/partitioned#L16)):*
> ```
>     PARTITION us VALUES IN (('earth', 'us', DEFAULT)),
>     PARTITION earth VALUES IN (('earth', DEFAULT)),
> 	  PARTITION mars VALUES IN (('mars', DEFAULT)),
> ```

this line has a tab but the others don't
___
*[pkg/sql/sem/tree/datum.go, line 158 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ8_Df1S8Auwestdot:-LkZ8_Dg6rvVEzppSV-N:b-m1u48p) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/sem/tree/datum.go#L158)):*
> ```Go
> 
> // Compare returns -1 if the receiver is less than other, 0 if receiver is
> // equal to other and +1 if receiver is greater than other.
> ```

mention lexicographical comparison
___
*[pkg/sql/sem/tree/datum.go, line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ8TuJDC0o6yaxpQXB:-LkZ8TuJDC0o6yaxpQXC:b-838czn) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/sem/tree/datum.go#L161)):*
> ```Go
> func (d Datums) Compare(evalCtx *EvalContext, other Datums) int {
> 	if len(d) == 0 {
> 		panic(errors.AssertionFailedf("empty datum being compared to other"))
> ```

datum -> Datums


<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
TFTR! They were incredibly thoughtful! I've addressed some of the reasons behind why I went this way in the comments and after talking to @RaduBerinde offline, going forward this way seems to be as good as the others in terms of perf. I also think its simpler this way and it is more consistent with the work done with the check constraints. This also involves just little additions to the index constraint library and doesn't require manipulating the spans outside the library at all. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-:-Lk_nDQ_1UMedjDst2dZ:bt5m7k5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj, @RaduBerinde, and @rytaft)

___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_constrained_scans, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkYh1qDCRpPMgRSfo_D:-Lk_WehEBUyMpTlckUc4:bk44dmh) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/ccl/logictestccl/testdata/logic_test/partitioning_constrained_scans#L1)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

should you add `fakedist-opt` or one of the others?
</blockquote></details>

Done. I didn't previously because the other partitioning tests didn't do this.
___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_constrained_scans, line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkWbsEH1DX5_kHMwYs5:-Lk_XvyqAyD1FtEJrXAh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/ccl/logictestccl/testdata/logic_test/partitioning_constrained_scans#L53)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

indentation seems off in this query
</blockquote></details>

Done.
___
*[pkg/sql/opt/constraint/constraint.go, line 81 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkWcIGp7UEz53i4EN1c:-Lk_YKPM3y0RWUR34Gxl:b6buv3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/constraint/constraint.go#L81)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This requires some clarification, and it needs unit tests (like `UnionWith` has). What does it mean exactly that they are not merged? It's one thing to not merge spans that "touch" like `[/a -  /b) [/b - /c]` and another thing to not merge spans that actually overlap (like `[/a - /c] [/b - /d]`. The spans in a Constraint are non-overlapping and if we violate that, a lot of code can break..
</blockquote></details>

Added unit tests and more elaboration. The code that uses this is guarantee to not have overlapping spans as they deal with the disjoint in-between spans (based on the partitions) but I added an assertion and comments to make this explicit. 
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 741 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkYypcY22ijJgeFLdD6:-Lk__E1LExJT0Uq1HMDy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/idxconstraint/index_constraints.go#L741)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] similar to `makeSpansForOr`
</blockquote></details>

Done.
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 780 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkYzPiKEi883T5B2ZAg:-Lk__FpsALlD3H2HPTsO:bv8w8ie) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/idxconstraint/index_constraints.go#L780)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Is this really necessary? What does the plan look like when they are merged?
</blockquote></details>

This is really useful sometimes. It keeps the likely spans separate and allow them to be constrained further.
Consider the the `solar_system` schema we use in the tests and the query `SELECT * FROM solar_system WHERE val = 8`. If we don't merge the spans that came from the partition filters we get:

` [ - /'earth'/'eu') [/'earth'/'eu' - /'earth'/'us'/'cali') [/'earth'/'us'/'cali' - /'earth'/'us'/'cali']...`

 Now consider the merged constraints that are equivalent:
 `[ - /'earth'/'us'/'cali']...`

Because of the given filter (`val = 8`),  notice how each of the spans above are further constrained:

 Unmerged: `[ - /'earth'/'eu') [/'earth'/'eu' - /'earth'/'us'/'cali') [/'earth'/'us'/'cali'/8 - /'earth'/'us'/'cali'/8]...`
 Merged:  `[ - /'earth'/'us'/'cali'/8]...`

In the unmerged case, we can skip all the values of the val col that have a value of less than 8 but have the prefix (earth, us, cali). This further constraining is another reason to keep them unmerged. If we merge them, we'll still use the index but will lose some of this ability, which is why I kept it this way.
___
*[pkg/sql/opt/ops/scalar.opt, line 311 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ6yw6ESzVMpbw6BTv:-Lk_eGQ7AWJ6cPF1bIBp:b-h0zbts) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/ops/scalar.opt#L311)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Having a separate type here for the sole purpose of not merging spans seems like a lot of (possibly unnecessary?) overhead
</blockquote></details>

As discussed in the unmerged comment, I think having the spans be unmerged is important. We could alternatively have them merge and just have a `Or` clause between all the `inPartition` and `inBetween` filters and this will guarantee that we will use the right index but it doesn't allow for further constraining to always happen. Radu's first suggestion to add the spans at the end does the same as this is doing now, except that we add the spans after the index constraint library has done it's thing. I thought it would be cleaner to add this logic here rather than later. I also like this better than manipulating the spans created for the suffix of the index, outside the index constraint library, because this way all the generation of spans happen in one place rather than multiple.

We also only use this when an index is not already constrained and when some filters are used that will further constrain the index. So in terms of overhead, I don't think its much different from either of the alternatives.
___
*[pkg/sql/opt/ops/scalar.opt, line 317 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ-1KpDNkh4rI2lq8L:-Lk_atNK1e4KioG_oYkI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/ops/scalar.opt#L317)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[super nit] kind of weird that this is on a new line but not a new paragraph.
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/custom_funcs.go, line 206 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ-yA4-dfoLIjdOTU9:-Lk_bqtp9eDaJfD6rYxI:b-uhid37) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/custom_funcs.go#L206)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] could reduce some clutter here by assigning the comparison expr to a variable in each case (since that's the only thing that changes), and then create the `FiltersItem` and append it to `filterExpr` below.
</blockquote></details>

Actually found a bug in the older logic here, we should've been doing tuple comparisons.
___
*[pkg/sql/opt/xform/custom_funcs.go, line 225 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ0jq6CvwKa1NqYWoP:-Lk_cgCYEpj7-pd_Dnrq:bv6redu) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/custom_funcs.go#L225)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Can you guarantee that this will never cause an array out of bounds error? Needs a comment explaining this.
</blockquote></details>

You were actually right! Found an out of bounds error in a test! Great catch!
___
*[pkg/sql/opt/xform/custom_funcs.go, line 237 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ1LpW44q02WwgO0Ne:-Lk_cns34_kWrmlc3bUT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/custom_funcs.go#L237)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] change this comma to semi-colon to fix the run-on sentence
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/custom_funcs.go, line 262 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ1pV8CWWLq6jWhnQG:-Lk_d3G47QPHezeKIat5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/custom_funcs.go#L262)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

you don't need this check since you'll just exit the loop below before any iterations
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/custom_funcs.go, line 311 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ5vn9Di5-a2k9q5vH:-Lk_d8OuD-Z9X1htN_Qr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/custom_funcs.go#L311)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] `partition` seems more appropriate than `partitionList`
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/custom_funcs.go, line 316 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ5Dmk2iPNTvtaGMf9:-Lk_dMLR6hJQA1Rkgis-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/custom_funcs.go#L316)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] don't love the naming here (`partitionSeen` v. `seenPartition`). Can you find a way to distinguish these better?
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/custom_funcs.go, line 508 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ7YVi48-5qLt_hwco:-Lk_dd1p-qsVTk9bLIKp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/custom_funcs.go#L508)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Update comment to mention partitioned index
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/testdata/rules/partitioned, line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ7pvY0k59tGT1-kUt:-Lk_fkuQ1BEmtJu2dIDQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/testdata/rules/partitioned#L7)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

what about adding another test with some regions/subregions in mars and/or Jupiter as well?
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/testdata/rules/partitioned, line 16 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ7i6o0kfqIhpgzlJZ:-Lk_dmEZ1TrsxCabnG-S:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/opt/xform/testdata/rules/partitioned#L16)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

this line has a tab but the others don't
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/datum.go, line 158 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ8_Df1S8Auwestdot:-Lk_e65O7PMpJqdMRUDI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/sem/tree/datum.go#L158)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

mention lexicographical comparison
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/datum.go, line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkZ8TuJDC0o6yaxpQXB:-Lk_eAu61ZKpbdwzpc5M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/773214e30c987551b332b5e0d4bc6533eaf7737d/pkg/sql/sem/tree/datum.go#L161)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

datum -> Datums
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
I need to think about this some more, but my initial impression is that going this direction will add cognitive overhead to every one of our interactions with the constraints libraries. They are already quite complex, so having another "mode" to think about will make them even harder to reason about. That said, I don't have an alternative (yet). Give this PR some time to bake in all our brains; perhaps someone will come up with a more constrained (ha!) change that will give us 80% of the benefits, but without adding as much to the abstraction complexity and maintenance cost.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-:-Lk_rGUT0-UUS8KDcAVp:b-fmkcrb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj, @RaduBerinde, and @rytaft)



<!-- Sent from Reviewable.io -->

ridwanmsharif commented:
After some discussion offline with @andy-kimball, and further reflection on whats possible here, I'm going to add a change that only adds the in between spans after the spans have been generated. This was @RaduBerinde 's original suggestion and it's roughly equivalent to what we're doing here. Except that it avoids the need to introduce another operator.

Commit:
opt: remove the use of the Likely expression

This change eliminates the need for the Likely operator
previously added. Instead we now constrain the index twice.
Once, using the partitions and again using the in between
spans. The constraints and remaining filters from both
of these are combined appropriately.

Release note: None

rytaft reviewed (commented):
I like the changes! I'll take another look after you fill in the comments, etc

> Reviewed 1 of 8 files at r3, 1 of 1 files at r4, 9 of 9 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-:-Lkcwv981N7bj7j31eTL:b-97juvh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj, @RaduBerinde, and @ridwanmsharif)

___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 1075 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkcuWGD2D1C3gz7NCDB:-LkcuWGD2D1C3gz7NCDC:b-2cqikg) ([raw file](https://github.com/cockroachdb/cockroach/blob/458b58a6695bc1b44c7b0ba8dd772a50eca0a41e/pkg/sql/opt/idxconstraint/index_constraints.go#L1075)):*
> ```Go
> func (ic *Instance) Init(
> 	filters memo.FiltersExpr,
> 	inBetween memo.FiltersExpr,
> ```

Need a comment about why a separate parameter is required here
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 1098 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-Lkctf0sAmZMvTDXQlwV:-Lkctf0t5MqFXcFABzep:b-qwrx9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/458b58a6695bc1b44c7b0ba8dd772a50eca0a41e/pkg/sql/opt/idxconstraint/index_constraints.go#L1098)):*
> ```Go
> 		if len(inBetween) > 0 {
> 			var exprConstraint constraint.Constraint
> 			inBetweenTight := ic.makeSpansForExpr(0, &inBetween, &exprConstraint)
> ```

[nit] `0 /* offset */`
___
*[pkg/sql/opt/xform/custom_funcs.go, line 350 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkcvXuNDPO2JEg5gH67:-LkcvXuNDPO2JEg5gH68:b-bg9oq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/458b58a6695bc1b44c7b0ba8dd772a50eca0a41e/pkg/sql/opt/xform/custom_funcs.go#L350)):*
> ```Go
> // TODO(ridwanmsharif): Comment on the changes made.
> func (c *CustomFuncs) constructOr(conditions memo.ScalarListExpr) opt.ScalarExpr {
> 	orExpr := c.e.f.ConstructOr(c.e.f.ConstructFalse(), c.e.f.ConstructFalse())
> ```

Instead of doing this, why don't you just add a check inside the loop if `i == 0`. Or just initialize `orExpr` with the first condition, and start iterating from `i=1`.
___
*[pkg/sql/opt/xform/custom_funcs.go, line 729 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkcwaUR8kR3J5x5GvUD:-LkcwaUR8kR3J5x5GvUE:blx7w3w) ([raw file](https://github.com/cockroachdb/cockroach/blob/458b58a6695bc1b44c7b0ba8dd772a50eca0a41e/pkg/sql/opt/xform/custom_funcs.go#L729)):*
> ```Go
> 
> // filterOuterCols returns all the outer cols of the filter expression.
> func (c *CustomFuncs) filterOuterCols(filters memo.FiltersExpr) opt.ColSet {
> ```

there's already a function that does this -- `CustomFuncs.FilterOuterCols`.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38963)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj, @RaduBerinde, and @ridwanmsharif)

___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 1073 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkdIll12x6b6eV4d6Nr:-LkdIll2-WbZxSv555dV:b-ynpr67) ([raw file](https://github.com/cockroachdb/cockroach/blob/458b58a6695bc1b44c7b0ba8dd772a50eca0a41e/pkg/sql/opt/idxconstraint/index_constraints.go#L1073)):*
> ```Go
> 
> // Init processes the filter and calculates the spans.
> func (ic *Instance) Init(
> ```

Why not add the in-between constraints in the calling code? I don't see a compelling reason to make this library API conceptually more complex.


<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38963)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 1073 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkdIll12x6b6eV4d6Nr:-LkdpBa54D2xKCxhSwaI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/458b58a6695bc1b44c7b0ba8dd772a50eca0a41e/pkg/sql/opt/idxconstraint/index_constraints.go#L1073)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Why not add the in-between constraints in the calling code? I don't see a compelling reason to make this library API conceptually more complex.
</blockquote></details>

Done.
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 1075 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkcuWGD2D1C3gz7NCDB:-LkdpCQlB4dqfm1hPbJG:b-7nte2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/458b58a6695bc1b44c7b0ba8dd772a50eca0a41e/pkg/sql/opt/idxconstraint/index_constraints.go#L1075)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Need a comment about why a separate parameter is required here
</blockquote></details>

Not needed anymore.
___
*[pkg/sql/opt/idxconstraint/index_constraints.go, line 1098 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-Lkctf0sAmZMvTDXQlwV:-LkdpGUq7T0IACV_C9Rt:b-7nte2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/458b58a6695bc1b44c7b0ba8dd772a50eca0a41e/pkg/sql/opt/idxconstraint/index_constraints.go#L1098)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] `0 /* offset */`
</blockquote></details>

Not needed anymore.
___
*[pkg/sql/opt/xform/custom_funcs.go, line 350 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkcvXuNDPO2JEg5gH67:-LkdpKVW9lzfdw7x6i7g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/458b58a6695bc1b44c7b0ba8dd772a50eca0a41e/pkg/sql/opt/xform/custom_funcs.go#L350)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Instead of doing this, why don't you just add a check inside the loop if `i == 0`. Or just initialize `orExpr` with the first condition, and start iterating from `i=1`.
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/custom_funcs.go, line 729 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkcwaUR8kR3J5x5GvUD:-LkdpLVqDYt4gI13a3oF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/458b58a6695bc1b44c7b0ba8dd772a50eca0a41e/pkg/sql/opt/xform/custom_funcs.go#L729)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

there's already a function that does this -- `CustomFuncs.FilterOuterCols`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> just a few more nits

> Reviewed 8 of 8 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-:-LkdtUIgAfY5YHvWuCc7:b6qkeu1)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, @RaduBerinde, and @ridwanmsharif)

___
*[pkg/sql/opt/memo/expr.go, line 208 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-Lkdr51U0Fvpn0uac9A1:-Lkdr51U0Fvpn0uac9A2:b-4jo8v5) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa18839385d40451b093ede43243cbe2c48eb612/pkg/sql/opt/memo/expr.go#L208)):*
> ```Go
> 
> // AddFilters adds items to n such that it contains all the FiltersItems
> // in other as well.
> ```

Mention that it removes duplicates. (And is this necessary? Otherwise you could just use the `CustomFuncs.ConcatFilters` function)
___
*[pkg/sql/opt/memo/expr.go, line 212 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkdqMzXCFfz0eBcxBMr:-LkdqMzXCFfz0eBcxBMs:b1jl3s1) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa18839385d40451b093ede43243cbe2c48eb612/pkg/sql/opt/memo/expr.go#L212)):*
> ```Go
> 	result := make(FiltersExpr, 0, len(*n)+len(other))
> 	for _, filter := range *n {
> 		result = append(result, filter)
> ```

I think you can just do `copy(result, *n)` (although you'll need to set the initial length of `result` to `len(*n)`)

Alternatively, why not just append directly to `n` (since you're assigning to `n` at the bottom anyway)? If you only want to search the first `len(*n)` results for duplicates, you can just save that number in a variable...
___
*[pkg/sql/opt/xform/custom_funcs.go, line 353 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkdrsU74hbFebTvKKuw:-LkdrsU74hbFebTvKKux:b-fr7yxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa18839385d40451b093ede43243cbe2c48eb612/pkg/sql/opt/xform/custom_funcs.go#L353)):*
> ```Go
> 	}
> 
> 	orExpr := c.e.f.ConstructOr(conditions[0], c.e.f.ConstructFalse())
> ```

why not just do `orExpr := conditions[0]`?


<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-:-LkeEyCB3FheJguvYh3O:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @justinj, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/memo/expr.go, line 208 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-Lkdr51U0Fvpn0uac9A1:-LkeF-XE6B8-0suTBtkK:b-lgofbs) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa18839385d40451b093ede43243cbe2c48eb612/pkg/sql/opt/memo/expr.go#L208)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Mention that it removes duplicates. (And is this necessary? Otherwise you could just use the `CustomFuncs.ConcatFilters` function)
</blockquote></details>

Using that now but we should still dedup the remaining filters, doing that now.
___
*[pkg/sql/opt/memo/expr.go, line 212 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkdqMzXCFfz0eBcxBMr:-LkeFCDd0GQjVFTZnsQC:b-hbbe5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa18839385d40451b093ede43243cbe2c48eb612/pkg/sql/opt/memo/expr.go#L212)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think you can just do `copy(result, *n)` (although you'll need to set the initial length of `result` to `len(*n)`)

Alternatively, why not just append directly to `n` (since you're assigning to `n` at the bottom anyway)? If you only want to search the first `len(*n)` results for duplicates, you can just save that number in a variable...
</blockquote></details>

Replaced this with dedup.
___
*[pkg/sql/opt/xform/custom_funcs.go, line 353 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkdrsU74hbFebTvKKuw:-LkeFFKr-lhhy6aGG3zN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa18839385d40451b093ede43243cbe2c48eb612/pkg/sql/opt/xform/custom_funcs.go#L353)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

why not just do `orExpr := conditions[0]`?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 8 of 8 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38963)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @justinj, @RaduBerinde, and @ridwanmsharif)

___
*[pkg/sql/opt/memo/expr.go, line 212 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkeIrFm9JrGQruzdB7W:-LkeIrFn50H2prOFOBcW:b-if7od2) ([raw file](https://github.com/cockroachdb/cockroach/blob/fffc3a10ebdd06e59845d6d48214f9d23f7165a9/pkg/sql/opt/memo/expr.go#L212)):*
> ```Go
> 
> 	// Only add it if it hasn't already been added.
> 	for i, filter := range *n {
> ```

Maybe faster to just sort (you can use `CustomFuncs.SortFilters`) and then dedup?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
I really like this newest approach!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-:-Lkeo3Px6NkfdN1TdGSh:bix81ox)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @justinj, @RaduBerinde, @ridwanmsharif, and @rytaft)

___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_constrained_scans, line 31 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-Lkeka-zDNhhRbycA5OV:-Lkeka-zDNhhRbycA5OW:b-kakf7s) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1205b11d2ce9252fec896c793a9d322c26a2d9/pkg/ccl/logictestccl/testdata/logic_test/partitioning_constrained_scans#L31)):*
> ```
> # Use the partition values to constrain the scan.
> query T
> EXPLAIN (OPT, VERBOSE) SELECT * FROM abc where c = 4
> ```

Would `EXPLAIN (OPT)` be sufficient here? I think it shows the index constraints. Costs and stats will change frequently, better to avoid when possible.
___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_constrained_scans, line 94 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-Lkel-Ky13J9q975sYA8:-Lkel-Ky13J9q975sYA9:b-m03ern) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1205b11d2ce9252fec896c793a9d322c26a2d9/pkg/ccl/logictestccl/testdata/logic_test/partitioning_constrained_scans#L94)):*
> ```
> # and so this is equivalent to an unconstrained scan.
> # TODO(ridwanmsharif): We can detect this case when deciding to use the partitioning
> # values by checking if a suffix of the index, not part of the partition values is used
> ```

"no part"? I don't know if this TODO belongs here, as it wouldn't affect the output of this test, would it?
___
*[pkg/sql/opt/xform/custom_funcs.go, line 200 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkelISM4QCOIYXTvZ4f:-LkelISM4QCOIYXTvZ4g:b-iqcw6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1205b11d2ce9252fec896c793a9d322c26a2d9/pkg/sql/opt/xform/custom_funcs.go#L200)):*
> ```Go
> ) memo.FiltersExpr {
> 	var filterExpr memo.FiltersExpr
> 	colTypes := make([]types.T, 0, len(values))
> ```

[nit] make it `len(values)` and do `colTypes[i] = ..` in the loop (good practice, append is slower)
___
*[pkg/sql/opt/xform/custom_funcs.go, line 205 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkelVOo7vmfXRSvhLfc:-LkelVOo7vmfXRSvhLfd:b4w3r9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1205b11d2ce9252fec896c793a9d322c26a2d9/pkg/sql/opt/xform/custom_funcs.go#L205)):*
> ```Go
> 	}
> 
> 	var columnVariables memo.ScalarListExpr
> ```

these can also be allocated to `len(values)`
___
*[pkg/sql/opt/xform/custom_funcs.go, line 224 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkemGr32HUaaDXQah_E:-LkemGr32HUaaDXQah_F:b-jusyn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1205b11d2ce9252fec896c793a9d322c26a2d9/pkg/sql/opt/xform/custom_funcs.go#L224)):*
> ```Go
> 	}
> 
> 	return filterExpr
> ```

This should just return a `FilterItem`
___
*[pkg/sql/opt/xform/custom_funcs.go, line 306 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkemUVhCdJ6-ShPROfj:-LkemUVhCdJ6-ShPROfk:bqdxz4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1205b11d2ce9252fec896c793a9d322c26a2d9/pkg/sql/opt/xform/custom_funcs.go#L306)):*
> ```Go
> 	tabID opt.TableID, index cat.Index, partitionValues []tree.Datums,
> ) memo.ScalarListExpr {
> 	var likelySatisfied memo.ScalarListExpr
> ```

Do we have to use `ScalarListExpr`? Wouldn't a `[]FiltersExpr` be more clear? (and then we don't have to put the slice headers on the heap)
___
*[pkg/sql/opt/xform/custom_funcs.go, line 331 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkenPznB6tCoxDKQg2j:-LkenPznB6tCoxDKQg2k:b-o4q2pq) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1205b11d2ce9252fec896c793a9d322c26a2d9/pkg/sql/opt/xform/custom_funcs.go#L331)):*
> ```Go
> 		}
> 
> 		// This partition has a more selective partition and so, will be taken care
> ```

"has a" -> "is a prefix of a"
___
*[pkg/sql/opt/xform/custom_funcs.go, line 355 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkenbK185J7wyX_Fa2n:-LkenbK185J7wyX_Fa2o:b-i248w5) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1205b11d2ce9252fec896c793a9d322c26a2d9/pkg/sql/opt/xform/custom_funcs.go#L355)):*
> ```Go
> 	orExpr := conditions[0]
> 	for i := 1; i < len(conditions); i++ {
> 		orExpr = c.e.f.ConstructOr(conditions[i], orExpr)
> ```

We are constructing `Or` with `FiltersExpr` as operands? That seems strange..

Is it the case that here the `FiltersExpr` will always have either one or two conditions? Perhaps it's worth creating a special data structure which stores at most two expressions, and pass around slices of those (and construct them as an `And` under the `Or`, not as a `FiltersExpr`)


<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-:-Lkf1x6i6YsLQvhU_iro:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @justinj, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/memo/expr.go, line 212 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkeIrFm9JrGQruzdB7W:-LkepFQN-pCEHUXvmaTc:b-9nm1wj) ([raw file](https://github.com/cockroachdb/cockroach/blob/fffc3a10ebdd06e59845d6d48214f9d23f7165a9/pkg/sql/opt/memo/expr.go#L212)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Maybe faster to just sort (you can use `CustomFuncs.SortFilters`) and then dedup?
</blockquote></details>

Done. I wanted to avoid the extra allocation so added an in place sort.
___
*[pkg/sql/opt/xform/custom_funcs.go, line 306 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkemUVhCdJ6-ShPROfj:-Lkf1UjuCtYysMCMJN7l:bd7nshm) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1205b11d2ce9252fec896c793a9d322c26a2d9/pkg/sql/opt/xform/custom_funcs.go#L306)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Do we have to use `ScalarListExpr`? Wouldn't a `[]FiltersExpr` be more clear? (and then we don't have to put the slice headers on the heap)
</blockquote></details>

Here we need a `ScalarListExpr` to accumulate all the expressions that bound the index columns by the partition values. This list of scalars will be `OR`ed together and returned at the end of the function.
___
*[pkg/sql/opt/xform/custom_funcs.go, line 355 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkenbK185J7wyX_Fa2n:-Lkeyu-EDRcRmgeuBYIl:bra78uh) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1205b11d2ce9252fec896c793a9d322c26a2d9/pkg/sql/opt/xform/custom_funcs.go#L355)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We are constructing `Or` with `FiltersExpr` as operands? That seems strange..

Is it the case that here the `FiltersExpr` will always have either one or two conditions? Perhaps it's worth creating a special data structure which stores at most two expressions, and pass around slices of those (and construct them as an `And` under the `Or`, not as a `FiltersExpr`)
</blockquote></details>

Good point. I changed `columnComparison` to return `AND` expressions (and just the scalar comparison expressions for the beginning and end spans). So here, its always either a scalar expr or an `AND` under the Or.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-:-LkfI_4u51DtpQWMxP7b:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @justinj, and @rytaft)



<!-- Sent from Reviewable.io -->

ridwanmsharif commented:
bors r+

rytaft commented:
___
*[pkg/sql/opt/memo/expr.go, line 212 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/38963#-LkeIrFm9JrGQruzdB7W:-Lki7PaGDtb9raQ8_7mY:b8n8tua) ([raw file](https://github.com/cockroachdb/cockroach/blob/fffc3a10ebdd06e59845d6d48214f9d23f7165a9/pkg/sql/opt/memo/expr.go#L212)):*
<details><summary><i>Previously, ridwanmsharif (Ridwan Sharif) wrote…</i></summary><blockquote>

Done. I wanted to avoid the extra allocation so added an in place sort.
</blockquote></details>

Makes sense, but it looks like you aren't taking advantage of the fact that the slice is sorted here. I was thinking you could call `Sort` inside this function, and then just do one O(n) pass through the sorted slice to remove duplicates. This algorithm is still O(n^2).


<!-- Sent from Reviewable.io -->
